File: 6-zigzag-conversion/6-zigzag-conversion.cs
public class Solution {
  public string Convert(string s, int numRows) {
    int n = s.Length;
    if (numRows == 1 || numRows >= n) return s;    
    StringBuilder sb = new StringBuilder();
    int step = 2 * numRows - 2;
    for (int i = 0; i < numRows; i++) {
      int i0 = i;
      int i1 = step - i;
      if (i%(numRows-1) == 0) i1 = n;
      while (i0 < n || i1 < n) {
        if (i0 < n) {
          sb.Append(s[i0]);
          i0 += step;
        }
        if (i1 < n) {
          sb.Append(s[i1]);
          i1 += step;
        }
      }
    }
    return sb.ToString();
  }
}

File: 6-zigzag-conversion/6-zigzag-conversion.scala
object Solution {
  def convert(s: String, numRows: Int): String = {
    val len = s.length
    if (len <= 1 || numRows <= 1 || len <= numRows) return s
    val sb = new StringBuilder()
    val step = 2*numRows - 2
    for (i <- 0 until numRows) {
      var index1 = i
      var index2 = if (i%(numRows-1) == 0) len else step - i
      while (index1 < len || index2 < len) {
        if (index1 < len) {
          sb += s.charAt(index1)
          index1 += step
        }
        if (index2 < len) {
          sb += s.charAt(index2)
          index2 += step
        }
      }
    }
    return sb.toString
  }
}